### ISSUE TITLE
Replace-By-Fee: Consequences on privacy and possible mitigations

### ISSUE THREAD

**Issue by LaurentMT** [OpenBitcoinPrivacyProject/wallet-ratings#36]
I would like to open a discussion about the consequences of Replace-By-Fee on privacy.

One of the main use cases envisioned for RBF is fee bumping which has obvious benefits in case of network congestion.

It's still early to say how wallets will implement this new feature but it's not  unlikely that many developers may chose the easiest and cheapest solution: a straightforward decrease of the change output in order to increase the fees. This solution introduces an obvious privacy leak which is easy to exploit with a custom listening node and which allows to identify the change output. Imho, that may be a serious problem if fee bumping becomes a very "popular" activity in the future.

Any thoughts on these matters ?

Which kind of solutions can we imagine (& recommend) in order to mitigate this risk ? 
Shower thought ideas: 
- wallet proposing to substract additional fees from payment output ? => requires an agreement between payer & payee
- wallet proposing to add an input and 2 outputs (returned to the payer but simulating an additional payment) ? => costly
- ...

**Comment by petertodd** [comment]
100% agree it's a privacy leak, but at least with full-RBF it's only a leak of what the change address is; with fss-RBF not only is it more expensive, but you're forced to add an additional input, further linking your coins together. In any case, I expect fee-bumping to be done somewhat infrequently, with fee estimation working most of the time.

RBF can also be used to pay multiple recipients in a row by adding new outputs - with full-RBF this is a very efficient way of adding outputs. Something wallets could do is occasionally split outputs when bumping fees, turning one output into two, making it unclear what is or is not change. Equally as txs spending an unconfirmed output reveals what output is change with high probability, splitting change outputs to pay additional recipients is at least no worse than the alternative for privacy.

**Comment by justusranvier** [comment]
This is another area where CPFP is safer than RBF.

**Comment by kristovatlas** [comment]
I'm not aware of any wallets currently looking to implement RBF, but this is definitely an important thing to keep track of. So far, it is one of the clearest examples I've seen of how passive blockchain observation can be strongly amplified by transaction broadcast network monitoring. We can be sure that many coin analytic agencies would increase their network monitoring if this were to be popularized, whereas most of them are not bothering right now.

**Comment by LaurentMT** [comment]
I fully agree with these comments. In the case of fee bumping, I also expect that RBF will be used as a second chance solution (the first one being fees estimation) and it's very early to estimate the potential impact.

But beyond this very specific case, I wonder if it makes sense for OBPP to extend its activity with a "proactive" approach: analysis of risks associated to proposals made by devs, proposition of mitigations before implementation by wallets/services...

On one hand, this kind of approach is surely not the most efficient for OBPP (time spent on proposals without any certainty about adoption and global impact). On the other hand, it might be more efficient for the whole ecosystem (avoiding the classic path "1st implementation => Detection of a privacy leak => 2nd implementation" repeated for all wallets/services).

As suggested by Kristov, a pragmatic approach may be the way to go: 
- first analysis limited to the most obvious risks,
- monitoring the adoption of new features in order to prioritize the subjects to analyze in details,
- ...

Just some random thoughts.

**Comment by petertodd** [comment]
@kristovatlas NikiWallet's maintainer is working on RBF-using features, as is GreenAddress.

@LaurentMT If the man power exists to do it, I'd certainly support OBPP looking at risks proactively, though I'd point out that this is something most of the Bitcoin Core developers are thinking about already. (e.g. one of the many objections raised to Hearn's GETUTXO's pull-req was the total lack of privacy implemented)

In the case of RBF you have the problem that there can be significant costs associated with achieving better privacy in certain situations; privacy becomes a trade-off between other goals. For instance, to increase the fee on a transaction the choices you have from highest to lowest cost are CPFP > FSS-RBF > full-RBF, while the privacy from lowest to highest is FSS-RBF < full-RBF < CPFP, with the difference between full-RBF and CPFP depending on how common it is for wallets to bump fees on incoming transactions; currently this is very uncommon, so full-RBF and CPFP have nearly the same privacy yet they're at opposite ends of the cost spectrum.

**Comment by kristovatlas** [comment]
new attack (network observer): identify the change output in a transactions by observing multiple versions of the same transaction adjusted via some RBF method.

**Comment by kristovatlas** [comment]
The attack for this issue is:

> Determine the change output of a transaction by observing different versions of it broadcast as part of an RBF process

The countermeasure relevant to this issue is:

> Avoid broadcasting more than one version of a transaction

There are no criteria currently listed under this countermeasure.

I propose this new criterion:

> Avoid broadcasting more than one version of transaction

This is grammatically styled after this criterion:

> II C 1 a: Avoids including addresses from multiple identity containers in the same balance query?

**Comment by kristovatlas** [comment]
Meeting discussion summary: You could do RBF where you add a new input to the updated tx and spend it all on fee, thus not leaking the position of change. Therefore, here's a new wording of the proposed criterion:

> Avoid broadcasting more than one version of transaction such that output amounts are modified in new versions.
